chore: started with migration of databases

SamTV12345 opened the pull request.

SamTV12345 committed:
chore: started with migration of databases

dvdsk commented:
I think the Db trait is perfect however I'm not sure we want to couple Microbin to sea-orm. I was thinking of having one module/folder called db with submodules/folders for each database implementation. So one for sea-orm and another for jsondb.

Another point, microbin is quite a simple app, do we need a relational database?

SamTV12345 committed:
chore: use directly sqlite pool and json database

SamTV12345 committed:
chore: removed all errors

SamTV12345 commented:
> I think the Db trait is perfect however I'm not sure we want to couple Microbin to sea-orm. I was thinking of having one module/folder called db with submodules/folders for each database implementation. So one for sea-orm and another for jsondb.
> 
> Another point, microbin is quite a simple app, do we need a relational database?

Yes you are right. I think we should not use an orm for such a simple table. I don't think that with the time we will scale so large that an orm is necessary. Since the last commit I reverted my changes and moved over to Database trait, rusqlite and also implemented the json database. So far looks okay but need to do some more testing and also write some tests. If you have time. Feel free to review. The code is now very clean and also performant without this global lock mutex of pastas.

SamTV12345 committed:
chore: added test

SamTV12345 committed:
chore: added tests for sqlite and jsondb

SamTV12345 committed:
chore: added on all pull requests

yara-blue commented on db/src/db/json/db.rs:
Please take a careful look at the current implementation of the JSON db. It will prevent data corruption/loss in case of power loss, this will not.

SamTV12345 replied:
Oh I see it first writes it to a temp file and then moves it over. Will adapt this solution

yara-blue commented on db/src/db/error.rs:
Clean solution! Though not how error handling is usually done in Rust. I would suggest using `color-eyre` (see: https://github.com/eyre-rs/eyre/tree/master) or `thiserror` instead.

yara-blue commented on src/args.rs:
```diff
@@ -144,6 +147,22 @@ pub struct Args {
     pub max_file_size_unencrypted_mb: usize,
 }
 
+impl From<Args> for DatabaseArgs {
```
The curse of backwards compatibility. Good solution!

SamTV12345 replied:
Thanks. Yeah thought why not use some language features to make it nice :)

yara-blue commented on src/main.rs:
```diff
@@ -108,8 +96,10 @@ async fn main() -> std::io::Result<()> {
+    let db_args = args.clone();
+    let db = get_database(db_args.into());
     let app_state = AppState {
-        pastas: Arc::new(Mutex::new(read_all(&args))),
```
Your arcing (wrapping it in an Arc) the database here but its already in a box. You can instead Arc it in get_database. (return Arc<dyn traits>).

SamTV12345 replied:
You're right. Box is kinda redundand if you have Arc.

yara-blue commented on db/src/db/mod.rs:
You can move this file outside the db directory and name it db.rs
That's a little more readable then having many tabs in your IDE all ending in mod.rs :)

yara-blue commented on db/src/db/json/mod.rs:
There is no need for this file. Could it be that you might misunderstand how mods work? If so try reading the section in the book about them https://doc.rust-lang.org/stable/book/ch07-05-separating-modules-into-different-files.html and or rust by example https://doc.rust-lang.org/stable/rust-by-example/mod.html

yara-blue reviewed (commented):
No time this week to fully review this (It is kinda huge). I've had a quick look and left some feedback. Great work btw! And thanks for setting up Github CI (I hate doing that so I am very grateful you spare me the pain).

One more general I dont understand: Why do we need PastaEntity? Could you explain why we dont 'just' serialize Pasta into sqlite as a value?

dvdsk commented:
Also let me again express my thanks for you maintaining Microbin with me :+1:, I feel like you are making great progress getting this codebase into good shape! I just wish I could keep up with al your work!

SamTV12345 committed:
chore: moved to new db retrieval

SamTV12345 committed:
chore: removed unneeded mod.rs files

SamTV12345 committed:
chore: fixed pipeline

SamTV12345 committed:
chore: arc directly database

SamTV12345 committed:
chore: use color_eyre for error reports

SamTV12345 commented:
> Also let me again express my thanks for you maintaining Microbin with me 👍, I feel like you are making great progress getting this codebase into good shape! I just wish I could keep up with al your work!

Sure. Microbin makes quite the fun and is something I can work outside of work on 😀. Thank you so much for all your reviews. I taught Rust myself and e.g. this module stuff that you don't need to paste everywhere mod.rs files was unknown to me.

yara-blue commented on db/src/lib.rs:
This is a unique way of doing mod, I haven't seen that before, surprised it works. 

What I would do is have:
```txt
- db.rs
- db
  |- json.rs
  |- sqlite.rs
- test_utils
```

then in db.rs you have:
```rust
mod test_utils;
mod sqlite:
mod json;
```

SamTV12345 replied:
Thanks for the review. I included this in the codebase.

yara-blue reviewed (changes requested):
Looked at everything :)

The mods are still a bit strange (see comment) and I still do not get why we have PastaEntity. Why can we just use the Pasta struct?

SamTV12345 committed:
chore: included feedback from review

SamTV12345 committed:
chore: included feedback from review
